docs/proposals/003-applications-outside-argocd-namespace.md

```diff
@@ -0,0 +1,418 @@
+
+* Provide a Git repository to the tenant for managing Argo CD `Application`
+  resources
+
+* Create an appropriate `AppProject` per tenant that restricts source to above
+  mentioned Git repository, restricts destination to above mentioned namespace
+  and restricts allowed resources to be `Application`.
+
+* Create an appropriate `Application` in Argo CD, which uses above mentionend
+  Git repository as source, and above mentioned Namespace as destination
+
+### Goals
+
+* Allow reconciliation from Argo CD `Application` resources from any namespace
+  in the cluster where the Argo CD control plane is installed to.
+
+* Allow declarative self-service management of `Application` resources from
+  users without access to the Argo CD control plane's namespace (e.g. `argocd`)
+
+* Be as less intrusive as possible, while not introducing additional controllers
+  or components.
+
+### Non-Goals
+
+* Allow reconciliation from Argo CD `Application` resources that are created in
+  remote clusters. We believe this would be possible with some adaptions, but
+  for this initial proposal we only consider the control plane's cluster as a
+  source for `Application` resources.
+
+* Allow users to create `AppProject` resources on their own. Since `AppProject`
+  is used as central entity for enforcing governance and security, these should
+  stay in the control of the cluster administrators.
+
+* Replace or modify Argo CD internal RBAC model
+
+## Proposal
+
+We suggest to adapt the current mechanisms of reconciliation of `Application`
+resources to include resources within the control plane's cluster, but outside
+the control plane's namespace (e.g. `argocd`).
+
+We think the following changes would need to be performed so that this will
+become possible in a secure manner:
+
+* Adapt the `AppProject` spec to include a new field, e.g. `sourceNamespaces`.
+  The values of this new field will define which applications will be allowed
+  to associate to the `AppProject`. Consider the following example:
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: AppProject
+  metadata:
+    name: some-project
+    namespace: argocd
+  spec:
+    sourceNamespaces:
+    - foo-ns
+    - bar-ns
+  ```
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: Application
+  metadata:
+    name: some-app
+    namespace: bar-ns
+  spec:
+    project: some-project
+  ```
+
+  ```yaml
+  apiVersion: argoproj.io/v1alpha1
+  kind: Application
+  metadata:
+    name: other-app
+    namespace: other-ns
+  spec:
+    project: some-project
+  ```
+
+  would allow `Application` resources that are created in either namespace
+  `foo-ns` or `bar-ns` to specify `some-project` in their `.spec.project`
+  field to associate themselves to the `AppProject` named `some-project`.
+  In the above example, the Application `some-app` would be allowed to associate
+  to the AppProject `some-project`, but the Application `other-app` would be
+  invalid.
```

Comment:
One thing that is not clear in this section is that how do you define invalid? In the implementation  details  you have mentioned aborting reconcile, but it is not clear to users and is only discoverable in the controller logs. Should it be rejecting via a validation webhook? or letting the creation to happen, and introduce a new status for Application as `Invalid` and maybe also have some events?

I think the former might be more clear and less error prone for users, especially as rejection can happen with a clear message, while the latter requires users to notice that the app status is invalid and search for the reason in the events or  in `.status.description`.


Also it is unclear in this proposal how updates to appProj is handled? I think we should immediately recheck all affected applications (all apps created in sourceNamespaces section of updated appproj) and update their status to invalid if required.


The downside of validationwebhook is that, when an app is created in a ns which initially was allowed in sourceNamespaces, we can not notify users that it is no longer valid, and the reconcile will only fail and might confuse them.

Reply:
> One thing that is not clear in this section is that how do you define invalid? In the implementation details you have mentioned aborting reconcile, but it is not clear to users and is only discoverable in the controller logs. 

This would be reflected in the `Application`'s `.status` field, similar to how a failed sync is reflected right now. Since it's the application controller who is updating the `.status` of an `Application`, it's also the application controller who will be enforcing whether an application will be reconciled or not. This check is happening constantly before reconciling any app, so when you remove a source namespace from the `AppProject`, it would not be allowed to sync anymore.